### PR TITLE
feat(mobile): new mobile UI

### PR DESCRIPTION
**Pull request by alextran1502**
This PR implements UI changes and improvements for the mobile app.

# UI changes

## Navigation bar

image

- Replace `Sharing` page with `Albums` page

## Search page

image

When navigating to the search page, the user is directed to the SearchInputPage where the search terms and filters can be entered. 

## Albums page

This page consolidates all remote album operations, including creating, sorting, searching, and quick filtering. 

image

## Ordering dropdown menu

The menu comes with a new UI and more relaxed spacing.

image

## Library page

The library page now includes quick access buttons to various views, including

- Favorite
- Archives
- Shared links
- Trash
- People
- Places
- Album on device
- Partners management
- Quick access to partner's library 

<p>
image

image
</p>

# Code changes

- Removed sharing page
- Removed shared album provider
- Consolidated all album's logic to its AlbumService
- Refactored album operations to follow the repository pattern UI > Provider > Service > Repository
- Simplified fetching of remote albums
- Reduced the trigger frequency of fetching remote album information, reduced lag when accessing the pages

### CODE REVIEW

File: mobile/lib/interfaces/album.interface.dart [hunk 1]
```diff
@@ -2,6 +2,7 @@ import 'package:immich_mobile/entities/album.entity.dart';
 import 'package:immich_mobile/entities/asset.entity.dart';
 import 'package:immich_mobile/entities/user.entity.dart';
 import 'package:immich_mobile/interfaces/database.interface.dart';
+import 'package:immich_mobile/pages/albums/albums.page.dart';
```
Comment (fyfrey) [inline mobile/lib/interfaces/album.interface.dart:5]: I think we don't want to import from pages in this layer. Can we move the QuickFterMode definition to either this file or a new 'model' file?

File: mobile/lib/entities/album.entity.dart [hunk 2]
```diff
@@ -114,7 +114,6 @@ class Album {
         ) &&
         shared == other.shared &&
         activityEnabled == other.activityEnabled &&
-        owner.value == other.owner.value &&
```
Comment (fyfrey) [inline mobile/lib/entities/album.entity.dart:117]: Why is this removed?
Do we instead need a special method to compare albums regardless of the owner?

File: mobile/lib/providers/album/album.provider.dart [hunk 3]
```diff
@@ -94,3 +136,31 @@ final albumRenderlistProvider =
   }
   return const Stream.empty();
 });
+
+class LocalAlbumsNotifier extends StateNotifier<List<Album>> {
+  LocalAlbumsNotifier(this.db) : super([]) {
+    final query = db.albums.filter().not().remoteIdIsNotNull();
```
Comment (fyfrey) [inline mobile/lib/providers/album/album.provider.dart:142]: Can we please not use a double negation?

File: mobile/lib/repositories/album.repository.dart [hunk 4]
```diff
@@ -118,4 +120,36 @@ class AlbumRepository extends DatabaseRepository implements IAlbumRepository {
...
   Future<void> deleteAllLocal() =>
       txn(() => db.albums.where().localIdIsNotNull().deleteAll());
+
+  @override
+  Future<List<Album>> search(
+      String searchTerm, QuickFilterMode filterMode) async {
+    switch (filterMode) {
+      case QuickFilterMode.all:
+        return await db.albums
+            .filter()
```
Comment (fyfrey) [inline mobile/lib/repositories/album.repository.dart:130]: You can simplify these cases by factoring out common parts (like filter, find all) and just modify the query in the switch cases

File: mobile/lib/services/album.service.dart [hunk 5]
```diff
@@ -16,6 +16,7 @@ import 'package:immich_mobile/entities/album.entity.dart';
 import 'package:immich_mobile/entities/asset.entity.dart';
 import 'package:immich_mobile/entities/store.entity.dart';
 import 'package:immich_mobile/entities/user.entity.dart';
+import 'package:immich_mobile/pages/albums/albums.page.dart';
```
Comment (fyfrey) [inline mobile/lib/services/album.service.dart:19]: Service should not import from pages

**Review by fyfrey — COMMENTED** [review]
Rework is looking good :)

File: mobile/lib/entities/album.entity.dart (hunk 2 above)
Comment (alextran1502) [reply mobile/lib/entities/album.entity.dart:117]: From my observation, the `IsarLink` here changes the hashcode of the object even if the value is the same. So when we create a `Set` of `Album`, they aren't correctly compared.

https://github.com/immich-app/immich/blob/5e8616b45cd8497dd52d1a2eeee5149c6f871738/mobile/lib/services/album.service.dart#L173

Comment (fyfrey) [reply mobile/lib/entities/album.entity.dart:117]: since this specific case is just server albums and we don't want changes to asset.entity.dart break this: we can just specify to use the remoteId for the set operations (dart sets allow to use custom hash/equals operations)

File: mobile/lib/providers/album/album.provider.dart (hunk 3 above)
Comment (fyfrey) [reply mobile/lib/providers/album/album.provider.dart:142]: alright, I was just hinting at doing `db.albums.where().remoteIdIsNull()` instead :)

File: mobile/lib/entities/album.entity.dart (hunk 2 above)
Comment (alextran1502) [reply mobile/lib/entities/album.entity.dart:117]: Do you know the syntax for this?

Comment (shenlong-tanwen) [reply mobile/lib/entities/album.entity.dart:117]: > Do you know the syntax for this?

You can probably use a [SplayTreeSet](https://api.flutter.dev/flutter/dart-collection/SplayTreeSet-class.html)

Comment (fyfrey) [reply mobile/lib/entities/album.entity.dart:117]: Far simpler:
https://api.flutter.dev/flutter/dart-collection/HashSet/HashSet.html
You can just set a custom compare and hash code operation in the set constructor

Comment (alextran1502) [reply mobile/lib/entities/album.entity.dart]: Thanks @fyfrey for a cool knowledge. The HashSet is now implemented. Please give it another round of review and approve if all look good

**Review by jrasm91 — APPROVED** [review]

File: mobile/lib/providers/album/album.provider.dart [hunk 6]
```diff
@@ -25,14 +25,22 @@ class AlbumNotifier extends StateNotifier<List<Album>> {
   }
 
   final AlbumService _albumService;
+  final Isar db;
```
Comment (fyfrey) [inline mobile/lib/providers/album/album.provider.dart:28]: please remove this instance variable (not needed/accessed anywhere, yet available externally!)
just keep the constructor as it was

File: mobile/lib/providers/album/album.provider.dart [hunk 7]
```diff
@@ -94,3 +147,31 @@ final albumRenderlistProvider =
...
+    query.findAll().then((value) {
+      if (mounted) {
+        state = value;
+      }
+    });
+
+    _streamSub = query.watch().listen((data) => state = data);
+  }
+
+  final Isar db;
```
Comment (fyfrey) [inline mobile/lib/providers/album/album.provider.dart:164]: same as above: Isar is never used except in the constructor. yet this makes it externally available

**Review by fyfrey — APPROVED** [review]

File: mobile/lib/providers/album/album.provider.dart (hunk 7 above)
Comment (alextran1502) [reply mobile/lib/providers/album/album.provider.dart:164]: You need to define it in the class instance to use it in the constructor, right?

image

**Comment by sunxiang0918** [comment]
How can I search LivePhotos on the new interface? Previously there was a separate button on the search interface.
